File: src/notifications.rs
use crate::{engine::App, store::notifications::NotificationDelivery};
use anyhow::{Context, Result, bail, ensure};
use chrono::Utc;
use serde::Serialize;
use sha2::{Digest, Sha256};
use std::time::Duration;
use tokio::task::JoinHandle;
use tokio_util::sync::CancellationToken;

pub const WEBHOOK_ENV: &str = "OCTOMUS_NOTIFICATION_WEBHOOK_URL";
const MAX_URL_BYTES: usize = 8192;
const MAX_PAYLOAD_BYTES: usize = 8192;
const MAX_REPOSITORY_BYTES: usize = 256;
const MAX_ID_BYTES: usize = 128;

struct WebhookDestination {
    url: reqwest::Url,
    id: String,
}
impl WebhookDestination {
    fn parse(raw: &str) -> Result<Self> {
        let raw = raw.trim();
        ensure!(
            !raw.is_empty() && raw.len() <= MAX_URL_BYTES,
            "Notification webhook URL is empty or exceeds the 8192-byte limit"
        );
        let url =
            reqwest::Url::parse(raw).context("Notification webhook URL is not a valid URL")?;
        ensure!(
            url.username().is_empty() && url.password().is_none(),
            "Notification webhook URL must not contain credentials"
        );
        ensure!(
            url.fragment().is_none(),
            "Notification webhook URL must not contain a fragment"
        );
        let host = url
            .host()
            .context("Notification webhook URL must contain a host")?;
        match url.scheme() {
            "https" => {}
            "http" => ensure!(
                match host {
                    url::Host::Ipv4(ip) => ip.is_loopback(),
                    url::Host::Ipv6(ip) => ip.is_loopback(),
                    url::Host::Domain(_) => false,
                },
                "Plain HTTP notification webhooks require a loopback IP address"
            ),
            _ => bail!("Notification webhook URL must use https, or http for a loopback IP"),
        }
        Ok(Self {
            id: format!("{:x}", Sha256::digest(url.as_str().as_bytes())),
            url,
        })
    }
}

#[derive(Serialize)]
pub struct AttentionEvent {
    pub schema_version: u32,
    pub event_id: String,
    pub occurred_at: String,
    pub repository: String,
    pub cycle_id: Option<String>,
    pub run_id: Option<String>,
    pub task_id: Option<String>,
    pub category: String,
    pub action: String,
}

pub fn payload(delivery: &NotificationDelivery) -> Result<Vec<u8>> {
    let event = AttentionEvent {
        schema_version: 1,
        event_id: delivery.event_id.clone(),
        occurred_at: delivery.created_at.clone(),
        repository: delivery.repository.clone(),
        cycle_id: delivery.cycle_id.clone(),
        run_id: delivery.run_id.clone(),
        task_id: delivery.task_id.clone(),
        category: delivery.category.clone(),
        action: delivery.action.clone(),
    };
    ensure!(
        event.repository.len() <= MAX_REPOSITORY_BYTES
            && event.event_id.len() <= MAX_ID_BYTES
            && event.cycle_id.as_deref().unwrap_or("").len() <= MAX_ID_BYTES
            && event.run_id.as_deref().unwrap_or("").len() <= MAX_ID_BYTES
            && event.task_id.as_deref().unwrap_or("").len() <= MAX_ID_BYTES
            && event.category.len() <= MAX_ID_BYTES
            && event.action.len() <= MAX_ID_BYTES,
        "invalid_payload"
    );
    let bytes = serde_json::to_vec(&event).context("invalid_payload")?;
    ensure!(bytes.len() <= MAX_PAYLOAD_BYTES, "invalid_payload");
    Ok(bytes)
}

fn classify(status: u16) -> (&'static str, bool) {
    if status == 408 || status == 429 || status >= 500 {
        ("http_status", true)
    } else {
        ("http_status", false)
    }
}

async fn deliver(
    client: &reqwest::Client,
    destination: &WebhookDestination,
    delivery: &NotificationDelivery,
) -> Result<u16, &'static str> {
    let body = payload(delivery).map_err(|_| "invalid_payload")?;
    let response = client
        .post(destination.url.clone())
        .header("content-type", "application/json")
        .body(body)
        .send()
        .await
        .map_err(|error| {
            if error.is_timeout() {
                "timeout"
            } else {
                "transport_error"
            }
        })?;
    let status = response.status().as_u16();
    drop(response);
    Ok(status)
}

pub fn start(app: &App, configured_url: Option<String>) -> Result<JoinHandle<()>> {
    let (destination, state, error) = match configured_url
        .as_deref()
        .map(str::trim)
        .filter(|value| !value.is_empty())
    {
        None => (None, "disabled", None),
        Some(raw) => match WebhookDestination::parse(raw) {
            Ok(destination) => (Some(destination), "enabled", None),
            Err(error) => (None, "invalid", Some(error.to_string())),
        },
    };
    let client = destination.as_ref().map(|_| {
        reqwest::Client::builder()
            .redirect(reqwest::redirect::Policy::none())
            .no_proxy()
            .timeout(Duration::from_secs(10))
            .connect_timeout(Duration::from_secs(5))
            .build()
    });
    let (destination, state, error, client) = match client {
        Some(Ok(client)) => (destination, state, error, Some(client)),
        Some(Err(_)) => (
            None,
            "invalid",
            Some("Notification webhook client could not be configured".to_owned()),
            None,
        ),
        None => (destination, state, error, None),
    };
    app.store.configure_notifications(
        destination.as_ref().map(|d| d.id.as_str()),
        state,
        error.as_deref(),
    )?;
    let app = app.clone();
    let shutdown: CancellationToken = app.shutdown.child_token();
    Ok(tokio::spawn(async move {
        let Some((destination, client)) = destination.zip(client) else {
            shutdown.cancelled().await;
            return;
        };
        let mut interval = tokio::time::interval(Duration::from_secs(1));
        interval.set_missed_tick_behavior(tokio::time::MissedTickBehavior::Delay);
        loop {
            tokio::select! {
                _ = shutdown.cancelled() => break,
                _ = interval.tick() => {}
            }
            if shutdown.is_cancelled() {
                break;
            }
            let delivery = match app.store.claim_notification(&destination.id, Utc::now()) {
                Ok(delivery) => delivery,
                Err(_) => {
                    tracing::warn!("Notification outbox claim failed");
                    continue;
                }
            };
            let Some(delivery) = delivery else { continue };
            let outcome = tokio::select! {
                _ = shutdown.cancelled() => return,
                outcome = deliver(&client, &destination, &delivery) => outcome,
            };
            let result = match outcome {
                Err(category) => app.store.finish_notification_failure(
                    delivery.seq,
                    category,
                    None,
                    category != "invalid_payload",
                ),
                Ok(status) if (200..300).contains(&status) => app
                    .store
                    .finish_notification_delivered(delivery.seq, Utc::now()),
                Ok(status) => {
                    let (category, retryable) = classify(status);
                    app.store.finish_notification_failure(
                        delivery.seq,
                        category,
                        Some(status),
                        retryable,
                    )
                }
            };
            if result.is_err() {
                tracing::warn!("Notification delivery update failed");
            }
        }
    }))
}

File: internal/notifications/notifications.go
// Package notifications owns durable webhook delivery: the outbox rows the
// store's triggers enqueue become attention events POSTed to the operator's
// configured destination, retried on remote failure, never exposing the URL.
package notifications

import (
	"bytes"
	"context"
	"errors"
	"fmt"
	"io"
	"net"
	"net/http"
	"net/url"
	"os"
	"strings"
	"sync"
	"time"

	"github.com/tyk-swe/octomus-agent/internal/model"
	"github.com/tyk-swe/octomus-agent/internal/redact"
	"github.com/tyk-swe/octomus-agent/internal/store"
	"github.com/tyk-swe/octomus-agent/internal/wirejson"
)

// WebhookEnv names the notification destination variable; its value is a secret.
const WebhookEnv = store.WebhookEnv

const (
	maxPayloadBytes    = 8192
	maxRepositoryBytes = 256
	maxIDBytes         = 128
)

// Delivery categories recorded in the outbox beside a failed attempt and shown
// as the dashboard's last notification error. invalidPayload is ours and never
// retried; every other failure is a remote condition.
const (
	httpStatusCategory = "http_status"
	invalidPayload     = "invalid_payload"
	timeoutCategory    = "timeout"
	transportCategory  = "transport_error"
)

// attentionEvent is the version-1 webhook payload.
type attentionEvent struct {
	SchemaVersion uint32  `json:"schema_version"`
	EventID       string  `json:"event_id"`
	OccurredAt    string  `json:"occurred_at"`
	Repository    string  `json:"repository"`
	CycleID       *string `json:"cycle_id"`
	RunID         *string `json:"run_id"`
	TaskID        *string `json:"task_id"`
	Category      string  `json:"category"`
	Action        string  `json:"action"`
}

// payload renders one outbox row as the bounded JSON body. Sizes are enforced
// before encoding so an unbounded stored field fails as invalid_payload rather
// than producing an over-limit request.
func payload(delivery *store.NotificationDelivery) ([]byte, error) {
	event := attentionEvent{
		SchemaVersion: 1,
		EventID:       delivery.EventID,
		OccurredAt:    delivery.CreatedAt,
		Repository:    delivery.Repository,
		CycleID:       delivery.CycleID,
		RunID:         delivery.RunID,
		TaskID:        delivery.TaskID,
		Category:      delivery.Category,
		Action:        delivery.Action,
	}
	if len(event.Repository) > maxRepositoryBytes ||
		len(event.EventID) > maxIDBytes ||
		len(deref(event.CycleID)) > maxIDBytes ||
		len(deref(event.RunID)) > maxIDBytes ||
		len(deref(event.TaskID)) > maxIDBytes ||
		len(event.Category) > maxIDBytes ||
		len(event.Action) > maxIDBytes {
		return nil, errors.New(invalidPayload)
	}
	encoded, err := wirejson.Marshal(event)
	if err != nil {
		return nil, fmt.Errorf("%s: %w", invalidPayload, err)
	}
	if len(encoded) > maxPayloadBytes {
		return nil, errors.New(invalidPayload)
	}
	return encoded, nil
}

func deref(value *string) string {
	if value == nil {
		return ""
	}
	return *value
}

// retryable reports whether a remote HTTP status deserves another attempt.
func retryable(status uint16) bool {
	return status == 408 || status == 429 || status >= 500
}

// Worker is the long-lived delivery loop. It owns no durable state: every
// outcome is written through the outbox operations.
type Worker struct {
	store    *store.Store
	url      string
	destID   string
	client   *http.Client
	ctx      context.Context
	cancel   context.CancelFunc
	done     chan struct{}
	shutdown sync.Once
	// warnings receives one redacted line per store failure episode; only the
	// run goroutine writes to it or reads and writes lastWarning.
	warnings    io.Writer
	lastWarning string
}

// webhookClient is the operator-safe delivery client: it follows no redirects
// (a 3xx is recorded as the delivery's HTTP status), ignores proxy environment
// variables, and bounds each request to 10 seconds and each connect to 5.
func webhookClient() *http.Client {
	transport := http.DefaultTransport.(*http.Transport).Clone()
	transport.Proxy = nil
	transport.DialContext = (&net.Dialer{Timeout: 5 * time.Second}).DialContext
	return &http.Client{
		Transport: transport,
		Timeout:   10 * time.Second,
		CheckRedirect: func(req *http.Request, via []*http.Request) error {
			return http.ErrUseLastResponse
		},
	}
}

// Start validates the configured webhook URL, records the durable policy and,
// when a destination is enabled, launches the delivery loop under the parent's
// shutdown scope. A nil worker is returned for disabled or invalid
// configuration — the policy write still happens so the dashboard reflects it.
// Store failures in the loop are reported on standard error.
func Start(parent context.Context, db *store.Store, configuredURL string) (*Worker, error) {
	return start(parent, db, configuredURL, os.Stderr)
}

func start(parent context.Context, db *store.Store, configuredURL string, warnings io.Writer) (*Worker, error) {
	raw := strings.TrimSpace(configuredURL)
	var normalized, destinationID string
	state := "disabled"
	var destination, errorText *string
	if raw != "" {
		destinationURL, id, err := model.NotificationDestination(raw)
		if err != nil {
			state = "invalid"
			message := err.Error()
			errorText = &message
		} else {
			normalized, destinationID = destinationURL, id
			state = "enabled"
			destination = &destinationID
		}
	}
	if err := db.ConfigureNotifications(destination, state, errorText); err != nil {
		return nil, err
	}
	if state != "enabled" {
		return nil, nil
	}
	ctx, cancel := context.WithCancel(parent)
	worker := &Worker{
		store:    db,
		url:      normalized,
		destID:   destinationID,
		client:   webhookClient(),
		ctx:      ctx,
		cancel:   cancel,
		done:     make(chan struct{}),
		warnings: warnings,
	}
	go worker.run()
	return worker, nil
}

// Stop cancels in-flight delivery and waits for the loop to exit. A claimed
// row abandoned mid-delivery stays claimed and retries on the next start.
func (w *Worker) Stop() {
	w.shutdown.Do(func() {
		w.cancel()
		<-w.done
	})
}

func (w *Worker) run() {
	defer close(w.done)
	// The delivery interval ticks immediately: a queued outbox row does not
	// wait a full second for its first attempt.
	timer := time.NewTimer(0)
	defer timer.Stop()
	for {
		select {
		case <-w.ctx.Done():
			return
		case <-timer.C:
		}
		timer.Reset(time.Second)
		if w.ctx.Err() != nil {
			return
		}
		delivery, err := w.store.ClaimNotification(w.destID, time.Now().UTC())
		if err != nil {
			w.warn(err)
			continue
		}
		// A working claim ends a failure episode: the next failure is reported
		// again even when its message repeats.
		w.lastWarning = ""
		if delivery == nil {
			continue
		}
		status, category := w.deliver(delivery)
		if w.ctx.Err() != nil {
			return
		}
		switch {
		case category != "":
			w.warn(w.store.FinishNotificationFailure(delivery.Seq, category, nil, category != invalidPayload))
		case status >= 200 && status < 300:
			w.warn(w.store.FinishNotificationDelivered(delivery.Seq, time.Now().UTC()))
		default:
			w.warn(w.store.FinishNotificationFailure(delivery.Seq, httpStatusCategory, &status, retryable(status)))
		}
	}
}

// warn reports a store failure once per episode: a loop stuck on the same
// failure every second writes one line, not one per tick. The message is
// redacted and never names the destination URL.
func (w *Worker) warn(err error) {
	if err == nil {
		return
	}
	message := redact.Error(err)
	if message == w.lastWarning {
		return
	}
	w.lastWarning = message
	fmt.Fprintf(w.warnings, "WARN notifications: %s\n", message)
}

// deliver posts one event; the category return names a local failure kind
// (invalidPayload, timeoutCategory or transportCategory) and otherwise the
// HTTP status decides.
func (w *Worker) deliver(delivery *store.NotificationDelivery) (uint16, string) {
	body, err := payload(delivery)
	if err != nil {
		return 0, invalidPayload
	}
	req, err := http.NewRequestWithContext(w.ctx, http.MethodPost, w.url, bytes.NewReader(body))
	if err != nil {
		return 0, transportCategory
	}
	req.Header.Set("content-type", "application/json")
	response, err := w.client.Do(req)
	if err != nil {
		if isTimeout(err) {
			return 0, timeoutCategory
		}
		return 0, transportCategory
	}
	defer response.Body.Close()
	return uint16(response.StatusCode), ""
}

// isTimeout reports whether a failed request timed out: request deadline
// expiry and transport timeouts count, but a caller cancellation does not.
func isTimeout(err error) bool {
	if errors.Is(err, context.DeadlineExceeded) {
		return true
	}
	var urlErr *url.Error
	if errors.As(err, &urlErr) {
		return urlErr.Timeout()
	}
	var netErr net.Error
	return errors.As(err, &netErr) && netErr.Timeout()
}
